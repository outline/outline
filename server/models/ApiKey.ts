import { Matches } from "class-validator";
import { subMinutes } from "date-fns";
import randomstring from "randomstring";
import { InferAttributes, InferCreationAttributes, Op } from "sequelize";
import {
  Column,
  Table,
  Unique,
  BeforeValidate,
  BelongsTo,
  ForeignKey,
  IsDate,
  DataType,
  AfterFind,
  BeforeSave,
} from "sequelize-typescript";
import { ApiKeyValidation } from "@shared/validations";
import { hash } from "@server/utils/crypto";
import User from "./User";
import ParanoidModel from "./base/ParanoidModel";
import { SkipChangeset } from "./decorators/Changeset";
import Fix from "./decorators/Fix";
import AuthenticationHelper from "./helpers/AuthenticationHelper";
import Length from "./validators/Length";

@Table({ tableName: "apiKeys", modelName: "apiKey" })
@Fix
class ApiKey extends ParanoidModel<
  InferAttributes<ApiKey>,
  Partial<InferCreationAttributes<ApiKey>>
> {
  static prefix = "ol_api_";

  static eventNamespace = "api_keys";

  /** The human-readable name of this API key */
  @Length({
    min: ApiKeyValidation.minNameLength,
    max: ApiKeyValidation.maxNameLength,
    msg: `Name must be between ${ApiKeyValidation.minNameLength} and ${ApiKeyValidation.maxNameLength} characters`,
  })
  @Column
  name: string;

  /** A list of scopes that this API key has access to */
  @Matches(/[\/\.\w\s]*/, {
    each: true,
  })
  @Column(DataType.ARRAY(DataType.STRING))
  scope: string[] | null;

  /** @deprecated The plain text value of the API key, removed soon. */
  @Unique
  @Column
  secret: string;

  /** The cached plain text value. Only available when creating the API key */
  @Column(DataType.VIRTUAL)
  value: string | null;

  /** The hashed value of the API key */
  @Unique
  @Column
  @SkipChangeset
  hash: string;

  /** The last 4 characters of the API key */
  @Column
  @SkipChangeset
  last4: string;

  /** The date and time when this API key will expire */
  @IsDate
  @Column
  expiresAt: Date | null;

  /** The date and time when this API key was last used */
  @IsDate
  @Column
  @SkipChangeset
  lastActiveAt: Date | null;

  // hooks

  @AfterFind
  public static async afterFindHook(models: ApiKey | ApiKey[]) {
    const modelsArray = Array.isArray(models) ? models : [models];
    for (const model of modelsArray) {
      if (model?.secret) {
        model.last4 = model.secret.slice(-4);
      }
    }
  }

  @BeforeValidate
  public static async generateSecret(model: ApiKey) {
    if (!model.hash) {
      const secret = `${ApiKey.prefix}${randomstring.generate(38)}`;
      model.value = model.secret || secret;
      model.hash = hash(model.value);
    }
  }

  @BeforeSave
  public static async updateLast4(model: ApiKey) {
    const value = model.value || model.secret;
    if (value) {
      model.last4 = value.slice(-4);
    }
  }

  /**
   * Validates that the input text _could_ be an API key, this does not check
   * that the key actually exists in the database.
   *
   * @param text The text to validate
   * @returns True if likely an API key
   */
  public static match(text: string) {
    // cannot guarantee prefix here as older keys do not include it.
    return !!text.replace(ApiKey.prefix, "").match(/^[\w]{38}$/);
  }

  /**
   * Finds an API key by the given input string. This will check both the
   * secret and hash fields.
   *
   * @param input The input string to search for
   * @returns The API key if found
   */
  public static findByToken(input: string) {
    return this.findOne({
      where: {
        [Op.or]: [{ secret: input }, { hash: hash(input) }],
      },
    });
  }

  // associations

  @BelongsTo(() => User, "userId")
  user: User;

  @ForeignKey(() => User)
  @Column
  userId: string;

  // methods

  updateActiveAt = async () => {
    const fiveMinutesAgo = subMinutes(new Date(), 5);

    // ensure this is updated only every few minutes otherwise
    // we'll be constantly writing to the DB as API requests happen
    if (!this.lastActiveAt || this.lastActiveAt < fiveMinutesAgo) {
      this.lastActiveAt = new Date();
    }

    return this.save({ silent: true });
  };

  /** Checks if the API key has access to the given path */
  canAccess = (path: string) => {
    if (!this.scope) {
      return true;
    }

    return AuthenticationHelper.canAccess(path, this.scope);
  };
}

export default ApiKey;
