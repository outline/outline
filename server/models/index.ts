export { default as ApiKey } from "./ApiKey";

export { default as Attachment } from "./Attachment";

export { default as AuthenticationProvider } from "./AuthenticationProvider";

export { default as Backlink } from "./Backlink";

export { default as Collection } from "./Collection";

export { default as GroupMembership } from "./GroupMembership";

export { default as UserMembership } from "./UserMembership";

export { default as Comment } from "./Comment";

export { default as Document } from "./Document";

export { default as Event } from "./Event";

export { default as FileOperation } from "./FileOperation";

export { default as Group } from "./Group";

export { default as GroupUser } from "./GroupUser";

export { default as Import } from "./Import";

export { default as ImportTask } from "./ImportTask";

export { default as Integration } from "./Integration";

export { default as IntegrationAuthentication } from "./IntegrationAuthentication";

export { default as Notification } from "./Notification";

export { default as Pin } from "./Pin";

export { default as Reaction } from "./Reaction";

export { default as Revision } from "./Revision";

export { default as SearchQuery } from "./SearchQuery";

export { default as Share } from "./Share";

export { default as Star } from "./Star";

export { default as Team } from "./Team";

export { default as TeamDomain } from "./TeamDomain";

export { default as User } from "./User";

export { default as UserAuthentication } from "./UserAuthentication";

export { default as View } from "./View";

export { default as WebhookSubscription } from "./WebhookSubscription";

export { default as WebhookDelivery } from "./WebhookDelivery";

export { default as Subscription } from "./Subscription";
