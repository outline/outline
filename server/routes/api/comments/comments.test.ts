import {
  CommentStatusFilter,
  ProsemirrorData,
  ReactionSummary,
} from "@shared/types";
import { Comment, Reaction } from "@server/models";
import {
  buildAdmin,
  buildCollection,
  buildComment,
  buildCommentMark,
  buildDocument,
  buildResolvedComment,
  buildTeam,
  buildUser,
} from "@server/test/factories";
import { getTestServer } from "@server/test/support";

const server = getTestServer();

describe("#comments.info", () => {
  it("should require authentication", async () => {
    const res = await server.post("/api/comments.info");
    const body = await res.json();
    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it("should return comment info", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const user2 = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });
    const comment = await buildComment({
      userId: user2.id,
      documentId: document.id,
    });
    const res = await server.post("/api/comments.info", {
      body: {
        token: user.getJwtToken(),
        id: comment.id,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.id).toEqual(comment.id);
    expect(body.data.data).toEqual(comment.data);
    expect(body.policies.length).toEqual(1);
    expect(body.policies[0].abilities.read).toBeTruthy();
    expect(body.policies[0].abilities.update).toEqual(false);
    expect(body.policies[0].abilities.delete).toEqual(false);
  });

  it("should return comment info for admin", async () => {
    const team = await buildTeam();
    const user = await buildAdmin({ teamId: team.id });
    const user2 = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });
    const comment = await buildComment({
      userId: user2.id,
      documentId: document.id,
    });
    const res = await server.post("/api/comments.info", {
      body: {
        token: user.getJwtToken(),
        id: comment.id,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.id).toEqual(comment.id);
    expect(body.data.data).toEqual(comment.data);
    expect(body.policies.length).toEqual(1);
    expect(body.policies[0].abilities.read).toBeTruthy();
    expect(body.policies[0].abilities.update).toBeTruthy();
    expect(body.policies[0].abilities.delete).toBeTruthy();
  });

  it("should return anchor text for an anchored comment", async () => {
    const anchorText = "anchor text";
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });
    const comment = await buildComment({
      userId: user.id,
      documentId: document.id,
    });
    const content = {
      type: "doc",
      content: [
        {
          type: "paragraph",
          content: [
            {
              type: "text",
              text: anchorText,
              marks: [buildCommentMark({ id: comment.id, userId: user.id })],
            },
          ],
        },
      ],
    } as ProsemirrorData;
    await document.update({ content });

    const res = await server.post("/api/comments.info", {
      body: {
        token: user.getJwtToken(),
        id: comment.id,
        includeAnchorText: true,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.id).toEqual(comment.id);
    expect(body.data.anchorText).toEqual(anchorText);
  });

  it("should not return anchor text for a non-anchored comment", async () => {
    const anchorText = "anchor text";
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });
    const comment = await buildComment({
      userId: user.id,
      documentId: document.id,
    });
    const content = {
      type: "doc",
      content: [
        {
          type: "paragraph",
          content: [
            {
              type: "text",
              text: anchorText,
              marks: [buildCommentMark({ userId: user.id })],
            },
          ],
        },
      ],
    } as ProsemirrorData;
    await document.update({ content });

    const res = await server.post("/api/comments.info", {
      body: {
        token: user.getJwtToken(),
        id: comment.id,
        includeAnchorText: true,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.id).toEqual(comment.id);
    expect(body.data.anchorText).toBeUndefined();
  });
});

describe("#comments.list", () => {
  it("should require authentication", async () => {
    const res = await server.post("/api/comments.list");
    const body = await res.json();
    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it("should return all comments for a document", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });
    const comment = await buildComment({
      userId: user.id,
      documentId: document.id,
    });
    await buildResolvedComment(user, {
      userId: user.id,
      documentId: document.id,
    });
    const res = await server.post("/api/comments.list", {
      body: {
        token: user.getJwtToken(),
        documentId: document.id,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.length).toEqual(2);
    expect(body.data[1].id).toEqual(comment.id);
    expect(body.policies.length).toEqual(2);
    expect(body.policies[0].abilities.read).toBeTruthy();
    expect(body.policies[1].abilities.read).toBeTruthy();
    expect(body.pagination.total).toEqual(2);
  });

  it("should return anchor texts for comments in a document", async () => {
    const anchorText = "anchor text";
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });
    const commentOne = await buildComment({
      userId: user.id,
      documentId: document.id,
    });
    const commentTwo = await buildResolvedComment(user, {
      userId: user.id,
      documentId: document.id,
    });
    const content = {
      type: "doc",
      content: [
        {
          type: "paragraph",
          content: [
            {
              type: "text",
              text: anchorText,
              marks: [buildCommentMark({ id: commentOne.id, userId: user.id })],
            },
          ],
        },
      ],
    } as ProsemirrorData;
    await document.update({ content });

    const res = await server.post("/api/comments.list", {
      body: {
        token: user.getJwtToken(),
        documentId: document.id,
        includeAnchorText: true,
        sort: "createdAt",
        direction: "ASC",
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.length).toEqual(2);
    expect(body.data[0].id).toEqual(commentOne.id);
    expect(body.data[1].id).toEqual(commentTwo.id);
    expect(body.data[0].anchorText).toEqual(anchorText);
    expect(body.data[1].anchorText).toBeUndefined();
  });

  it("should return unresolved comments for a collection", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const collection = await buildCollection({
      userId: user.id,
      teamId: team.id,
    });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
      collectionId: collection.id,
    });
    const comment = await buildComment({
      userId: user.id,
      documentId: document.id,
    });
    const res = await server.post("/api/comments.list", {
      body: {
        token: user.getJwtToken(),
        collectionId: collection.id,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.length).toEqual(1);
    expect(body.data[0].id).toEqual(comment.id);
    expect(body.policies.length).toEqual(1);
    expect(body.policies[0].abilities.read).toBeTruthy();
    expect(body.pagination.total).toEqual(1);
  });

  it("should return unresolved comments for a parentCommentId", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });
    const comment = await buildComment({
      userId: user.id,
      documentId: document.id,
    });
    const childComment = await buildComment({
      userId: user.id,
      documentId: document.id,
      parentCommentId: comment.id,
    });
    const res = await server.post("/api/comments.list", {
      body: {
        token: user.getJwtToken(),
        parentCommentId: comment.id,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.length).toEqual(1);
    expect(body.data[0].id).toEqual(childComment.id);
    expect(body.policies.length).toEqual(1);
    expect(body.policies[0].abilities.read).toBeTruthy();
    expect(body.pagination.total).toEqual(1);
  });

  it("should return resolved comments for a statusFilter", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });
    await buildComment({
      userId: user.id,
      documentId: document.id,
    });
    const resolved = await buildResolvedComment(user, {
      userId: user.id,
      documentId: document.id,
    });
    const res = await server.post("/api/comments.list", {
      body: {
        token: user.getJwtToken(),
        documentId: document.id,
        statusFilter: [CommentStatusFilter.Resolved],
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.length).toEqual(1);
    expect(body.data[0].id).toEqual(resolved.id);
    expect(body.policies.length).toEqual(1);
    expect(body.policies[0].abilities.read).toBeTruthy();
    expect(body.policies[0].abilities.unresolve).toBeTruthy();
    expect(body.policies[0].abilities.resolve).toEqual(false);
    expect(body.pagination.total).toEqual(1);
  });

  it("should return all unresolved comments", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const collection1 = await buildCollection({
      userId: user.id,
      teamId: team.id,
    });
    const collection2 = await buildCollection({
      userId: user.id,
      teamId: team.id,
    });
    const document1 = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
      collectionId: collection1.id,
    });
    const document2 = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
      collectionId: collection2.id,
    });
    const comment1 = await buildComment({
      userId: user.id,
      documentId: document1.id,
    });
    const comment2 = await buildComment({
      userId: user.id,
      documentId: document2.id,
    });
    const res = await server.post("/api/comments.list", {
      body: {
        token: user.getJwtToken(),
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.length).toEqual(2);
    expect([body.data[0].id, body.data[1].id].sort()).toEqual(
      [comment1.id, comment2.id].sort()
    );
    expect(body.policies.length).toEqual(2);
    expect(body.policies[0].abilities.read).toBeTruthy();
    expect(body.policies[1].abilities.read).toBeTruthy();
    expect(body.pagination.total).toEqual(2);
  });

  it("should return reactions for a comment", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });
    const reactions: ReactionSummary[] = [
      { emoji: "😄", userIds: [user.id] },
      { emoji: "🙃", userIds: [user.id] },
    ];
    const comment = await buildComment({
      userId: user.id,
      documentId: document.id,
      reactions,
    });

    const res = await server.post("/api/comments.list", {
      body: {
        token: user.getJwtToken(),
        documentId: document.id,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.length).toEqual(1);
    expect(body.data[0].id).toEqual(comment.id);
    expect(body.data[0].reactions).toEqual(reactions);
    expect(body.policies.length).toEqual(1);
    expect(body.policies[0].abilities.read).toBeTruthy();
    expect(body.pagination.total).toEqual(1);
  });
});

describe("#comments.create", () => {
  it("should require authentication", async () => {
    const res = await server.post("/api/comments.create");
    const body = await res.json();
    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it("should create a comment", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });

    const comment = await buildComment({
      userId: user.id,
      documentId: document.id,
    });

    const res = await server.post("/api/comments.create", {
      body: {
        token: user.getJwtToken(),
        documentId: document.id,
        data: comment.data,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.data).toEqual(comment.data);
    expect(body.policies.length).toEqual(1);
    expect(body.policies[0].abilities.read).toBeTruthy();
    expect(body.policies[0].abilities.update).toBeTruthy();
    expect(body.policies[0].abilities.delete).toBeTruthy();
  });

  it("should create a comment from markdown text", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });

    const text = "## heading\n\n- list item 1\n- list item 2";

    const res = await server.post("/api/comments.create", {
      body: {
        token: user.getJwtToken(),
        documentId: document.id,
        text,
      },
    });

    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.data).toMatchSnapshot();
  });

  it("should not allow empty comment data", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });

    const res = await server.post("/api/comments.create", {
      body: {
        token: user.getJwtToken(),
        documentId: document.id,
        data: null,
      },
    });

    const anotherRes = await server.post("/api/comments.create", {
      body: {
        token: user.getJwtToken(),
        documentId: document.id,
        data: {
          type: "doc",
          content: [{ type: "paragraph" }],
        },
      },
    });

    expect(res.status).toEqual(400);
    expect(anotherRes.status).toEqual(400);
  });

  it("should not allow comments containing only whitespaces", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });

    const res = await server.post("/api/comments.create", {
      body: {
        token: user.getJwtToken(),
        documentId: document.id,
        data: {
          type: "doc",
          content: [
            {
              type: "paragraph",
              content: [{ type: "text", text: "  \n\r\n" }],
            },
          ],
        },
      },
    });

    expect(res.status).toEqual(400);
  });

  it("should allow adding images to comments", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });

    const res = await server.post("/api/comments.create", {
      body: {
        token: user.getJwtToken(),
        documentId: document.id,
        data: {
          type: "doc",
          content: [
            {
              type: "paragraph",
              content: [
                {
                  type: "image",
                  attrs: {
                    src: "https://example.com/image.png",
                    alt: "Example image",
                  },
                },
              ],
            },
          ],
        },
      },
    });

    expect(res.status).toEqual(200);
  });

  it("should allow adding images from internal sources", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });

    const res = await server.post("/api/comments.create", {
      body: {
        token: user.getJwtToken(),
        documentId: document.id,
        data: {
          type: "doc",
          content: [
            {
              type: "paragraph",
              content: [
                {
                  type: "image",
                  attrs: {
                    src: "/api/attachments.redirect?id=1401323b-c4e2-40de-b172-e1668ec89111",
                    alt: null,
                  },
                },
              ],
            },
          ],
        },
      },
    });

    expect(res.status).toEqual(200);
  });

  it("should not allow invalid comment data", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });

    const res = await server.post("/api/comments.create", {
      body: {
        token: user.getJwtToken(),
        documentId: document.id,
        data: {
          type: "nonsense",
        },
      },
    });

    expect(res.status).toEqual(400);
  });
});

describe("#comments.update", () => {
  it("should require authentication", async () => {
    const res = await server.post("/api/comments.update");
    const body = await res.json();
    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it("should update an existing comment", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });

    const comment = await buildComment({
      userId: user.id,
      documentId: document.id,
    });

    const res = await server.post("/api/comments.update", {
      body: {
        token: user.getJwtToken(),
        id: comment.id,
        data: comment.data,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.data).toEqual(comment.data);
    expect(body.policies.length).toEqual(1);
    expect(body.policies[0].abilities.read).toBeTruthy();
    expect(body.policies[0].abilities.update).toBeTruthy();
    expect(body.policies[0].abilities.delete).toBeTruthy();
  });
});

describe("#comments.resolve", () => {
  it("should require authentication", async () => {
    const res = await server.post("/api/comments.resolve");
    const body = await res.json();
    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it("should allow resolving a comment thread", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });

    const comment = await buildComment({
      userId: user.id,
      documentId: document.id,
    });

    const res = await server.post("/api/comments.resolve", {
      body: {
        token: user.getJwtToken(),
        id: comment.id,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.updatedAt).toEqual(comment.updatedAt.toISOString());
    expect(body.data.resolvedAt).toBeTruthy();
    expect(body.data.resolvedById).toEqual(user.id);
    expect(body.data.resolvedBy.id).toEqual(user.id);
    expect(body.policies.length).toEqual(1);
    expect(body.policies[0].abilities.read).toBeTruthy();
    expect(body.policies[0].abilities.update).toBeTruthy();
    expect(body.policies[0].abilities.delete).toBeTruthy();
    expect(body.policies[0].abilities.unresolve).toBeTruthy();
    expect(body.policies[0].abilities.resolve).toEqual(false);
  });

  it("should not allow resolving a child comment", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });

    const parentComment = await buildComment({
      userId: user.id,
      documentId: document.id,
    });

    const comment = await buildComment({
      userId: user.id,
      documentId: document.id,
      parentCommentId: parentComment.id,
    });

    const res = await server.post("/api/comments.resolve", {
      body: {
        token: user.getJwtToken(),
        id: comment.id,
      },
    });
    expect(res.status).toEqual(403);
  });
});

describe("#comments.unresolve", () => {
  it("should require authentication", async () => {
    const res = await server.post("/api/comments.unresolve");
    const body = await res.json();
    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it("should allow unresolving a comment", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });

    const comment = await buildResolvedComment(user, {
      userId: user.id,
      documentId: document.id,
    });

    const res = await server.post("/api/comments.unresolve", {
      body: {
        token: user.getJwtToken(),
        id: comment.id,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.updatedAt).toEqual(comment.updatedAt.toISOString());
    expect(body.data.resolvedAt).toEqual(null);
    expect(body.data.resolvedBy).toEqual(null);
    expect(body.data.resolvedById).toEqual(null);
    expect(body.policies.length).toEqual(1);
    expect(body.policies[0].abilities.read).toBeTruthy();
    expect(body.policies[0].abilities.update).toBeTruthy();
    expect(body.policies[0].abilities.delete).toBeTruthy();
    expect(body.policies[0].abilities.resolve).toBeTruthy();
    expect(body.policies[0].abilities.unresolve).toEqual(false);
  });
});

describe("#comments.add_reaction", () => {
  it("should require authentication", async () => {
    const res = await server.post("/api/comments.add_reaction");
    const body = await res.json();
    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it("should add a reaction to a comment", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });
    const comment = await buildComment({
      userId: user.id,
      documentId: document.id,
    });

    const res = await server.post("/api/comments.add_reaction", {
      body: {
        token: user.getJwtToken(),
        id: comment.id,
        emoji: "😄",
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.success).toEqual(true);

    const updatedComment = await Comment.findByPk(comment.id);
    const addedReaction = await Reaction.findOne({
      where: { commentId: comment.id, emoji: "😄", userId: user.id },
    });

    expect(updatedComment?.reactions).toEqual([
      { emoji: "😄", userIds: [user.id] },
    ]);
    expect(addedReaction).toBeTruthy();
  });

  it("should add a reaction to a comment with existing reactions", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });
    const comment = await buildComment({
      userId: user.id,
      documentId: document.id,
      reactions: [{ emoji: "😄", userIds: ["test-user"] }],
    });

    const res = await server.post("/api/comments.add_reaction", {
      body: {
        token: user.getJwtToken(),
        id: comment.id,
        emoji: "😄",
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.success).toEqual(true);

    const updatedComment = await Comment.findByPk(comment.id);
    const addedReaction = await Reaction.findOne({
      where: { commentId: comment.id, emoji: "😄", userId: user.id },
    });

    expect(updatedComment?.reactions).toEqual([
      { emoji: "😄", userIds: ["test-user", user.id] },
    ]);
    expect(addedReaction).toBeTruthy();
  });
});

describe("#comments.remove_reaction", () => {
  it("should require authentication", async () => {
    const res = await server.post("/api/comments.remove_reaction");
    const body = await res.json();
    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it("should remove a reaction from a comment", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });
    const comment = await buildComment({
      userId: user.id,
      documentId: document.id,
    });
    await Reaction.create(
      {
        emoji: "😄",
        commentId: comment.id,
        userId: user.id,
      },
      { hooks: false }
    );

    const res = await server.post("/api/comments.remove_reaction", {
      body: {
        token: user.getJwtToken(),
        id: comment.id,
        emoji: "😄",
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.success).toEqual(true);

    const updatedComment = await Comment.findByPk(comment.id);
    const removedReaction = await Reaction.findOne({
      where: { commentId: comment.id, emoji: "😄", userId: user.id },
    });

    expect(updatedComment?.reactions).toEqual([]);
    expect(removedReaction).toBeNull();
  });

  it("should remove a reaction from a comment with existing reactions", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const document = await buildDocument({
      userId: user.id,
      teamId: user.teamId,
    });
    const comment = await buildComment({
      userId: user.id,
      documentId: document.id,
      reactions: [{ emoji: "😄", userIds: ["test-user"] }],
    });
    await Reaction.create(
      {
        emoji: "😄",
        commentId: comment.id,
        userId: user.id,
      },
      { hooks: false }
    );

    const res = await server.post("/api/comments.remove_reaction", {
      body: {
        token: user.getJwtToken(),
        id: comment.id,
        emoji: "😄",
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.success).toEqual(true);

    const updatedComment = await Comment.findByPk(comment.id);
    const removedReaction = await Reaction.findOne({
      where: { commentId: comment.id, emoji: "😄", userId: user.id },
    });

    expect(updatedComment?.reactions).toEqual([
      { emoji: "😄", userIds: ["test-user"] },
    ]);
    expect(removedReaction).toBeNull();
  });
});
