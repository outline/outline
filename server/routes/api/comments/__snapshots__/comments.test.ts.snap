// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#comments.add_reaction should require authentication 1`] = `
{
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#comments.create should create a comment from markdown text 1`] = `
{
  "content": [
    {
      "attrs": {
        "level": 2,
      },
      "content": [
        {
          "text": "heading",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    {
      "content": [
        {
          "content": [
            {
              "content": [
                {
                  "text": "list item 1",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "list_item",
        },
        {
          "content": [
            {
              "content": [
                {
                  "text": "list item 2",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "list_item",
        },
      ],
      "type": "bullet_list",
    },
  ],
  "type": "doc",
}
`;

exports[`#comments.create should require authentication 1`] = `
{
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#comments.info should require authentication 1`] = `
{
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#comments.list should require authentication 1`] = `
{
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#comments.remove_reaction should require authentication 1`] = `
{
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#comments.resolve should require authentication 1`] = `
{
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#comments.unresolve should require authentication 1`] = `
{
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#comments.update should require authentication 1`] = `
{
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;
